mmv1/third_party/terraform/utils/kms_utils.go

```diff
@@ -188,16 +220,37 @@ func clearCryptoKeyVersions(cryptoKeyId *kmsCryptoKeyId, userAgent string, confi
 	}
 
 	for _, version := range versionsResponse.CryptoKeyVersions {
-		request := &cloudkms.DestroyCryptoKeyVersionRequest{}
-		destroyCall := versionsClient.Destroy(version.Name, request)
-		if config.UserProjectOverride {
-			destroyCall.Header().Set("X-Goog-User-Project", cryptoKeyId.KeyRingId.Project)
+		// skip the versions that have been destroyed earlier
+		if version.State == "ENABLED" {
+			request := &cloudkms.DestroyCryptoKeyVersionRequest{}
+			destroyCall := versionsClient.Destroy(version.Name, request)
+			if config.UserProjectOverride {
+				destroyCall.Header().Set("X-Goog-User-Project", cryptoKeyId.KeyRingId.Project)
+			}
+			_, err = destroyCall.Do()
+
+			if err != nil {
+				return err
+			}
 		}
-		_, err = destroyCall.Do()
+	}
 
-		if err != nil {
-			return err
-		}
+	return nil
+}
+
+func deleteCryptoKeyVersions(cryptoKeyVersionId *kmsCryptoKeyVersionId, d *schema.ResourceData, userAgent string, config *Config) error {
+	versionsClient := config.NewKmsClient(userAgent).Projects.Locations.KeyRings.CryptoKeys.CryptoKeyVersions
+	request := &cloudkms.DestroyCryptoKeyVersionRequest{}
+	destroyCall := versionsClient.Destroy(cryptoKeyVersionId.Name, request)
+	if config.UserProjectOverride {
+		destroyCall.Header().Set("X-Goog-User-Project", cryptoKeyVersionId.CryptoKeyId.KeyRingId.Project)
+	}
+	response, err := destroyCall.Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("ID %s", cryptoKeyVersionId.Name))
+	}
+	if response.State == "DESTROY_SCHEDULED" {
```

Comment:
it looks like this is sufficient to not throw an error but I'm not sure it has the desired effect, since the resource will not be removed from state.

If the response for DESTROY_SCHEDULED is a 404 error it should go through the path above. Do you know whether this path is ever actually hit? If not - it would be safe to remove. If so - it might need custom handling and we need to figure out why the tests are still passing. We might need to add a custom check to make sure that the resource is actually deleted if the builtin behavior can't detect it... https://developer.hashicorp.com/terraform/plugin/sdkv2/testing/acceptance-tests/teststep#custom-check-functions

Reply:
This line was hit before, but it does not throw the proper error out. I think adding a custom handling make sense here.